src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -0,0 +1,288 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
+{
+    internal class SemanticTokensHelpers
+    {
+        /// <summary>
+        /// Returns the semantic tokens for a given document with an optional range.
+        /// </summary>
+        internal static async Task<LSP.SemanticTokens> ComputeSemanticTokensAsync(
+            TextDocumentIdentifier? textDocument,
+            int previousResultId,
+            string? clientName,
+            ILspSolutionProvider solutionProvider,
+            LSP.Range? range,
+            CancellationToken cancellationToken)
+        {
+            if (textDocument == null)
+            {
+                return new LSP.SemanticTokens();
+            }
+
+            var document = solutionProvider.GetDocument(textDocument, clientName);
+            if (document == null)
+            {
+                return new LSP.SemanticTokens();
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            if (root == null)
+            {
+                return new LSP.SemanticTokens();
+            }
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            if (text == null)
+            {
+                return new LSP.SemanticTokens();
+            }
+
+            var textSpan = root.FullSpan;
+            if (range != null)
+            {
+                textSpan = ProtocolConversions.RangeToTextSpan(range, text);
+            }
+
+            var classifiedSpans = await Classifier.GetClassifiedSpansAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (classifiedSpans == null)
+            {
+                return new LSP.SemanticTokens();
+            }
+
+            var lastLineNumber = 0;
+            var lastStartCharacter = 0;
+
+            var groupedSpans = classifiedSpans.GroupBy(s => s.TextSpan);
```

Comment:
feels very weird that we're grouping anything.

Reply:
I added a clarifying comment in the latest updates that hopefully explains things more.
The GroupBy behavior is modeled after how the Omnisharp SemanticHighlightService works - https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Roslyn.CSharp/Services/SemanticHighlight/SemanticHighlightService.cs#L53-L65

Reply:
yeah, i would prefer we not do this.  Classifier.GetClassifiedSpansAsync will already report an ordered list of classification spans.  so you just need to handle the case of getting multiple spans in a row with teh same text span.

Reply:
this comment applies.  the tokens will be in order already.  so i would just linearly process them.

Reply:
note: the above is *very* expensive.  you're basically allocating multiple times *per* classification.

Reply:
we still need to address thsi.  classified spans are sorted.  tehre is no need to group anything.  if we want to handle teh case of `(0, 5, identifier),(0, 5, static)` we can do so as we linearly traverse this list looking to see if multiple classified spans in a row refer to the same text span.

this is a blocker for me.

Reply:
I agree with Cyrus. For OmniSharp the semantic feature in opt-in and we tell customers it is experimental.

Since these spans are in order, looping through them, processing the ClassificationType, and emitting a Token when the span changes would perform better.